smack-extensions/src/main/java/org/jivesoftware/smackx/usertune/element/UserTuneElement.java

```diff
@@ -0,0 +1,223 @@
+        if (isNullUserTune()) {
+            return xml.closeEmptyElement();
+        }
+        xml.rightAngleBracket();
+        xml.optAppend(getUserTuneXML());
+        return xml.closeElement(getElementName());
+    }
+
+    private CharSequence getUserTuneXML() {
+
+        XmlStringBuilder userTuneXML = new XmlStringBuilder();
+            if (artist != null) {
+                userTuneXML.element("artist", artist);
+            }
+            if (length  != -1) {
+                userTuneXML.element("length", Integer.toString(length));
+            }
+            if (rating  != -1) {
+                userTuneXML.element("rating",Integer.toString(rating));
+            }
+            userTuneXML.element("source",source);
+            userTuneXML.element("title",title);
+            if (track != -1) {
+                userTuneXML.element("track",Integer.toString(track));
+            }
+            if (uri != null) {
+                userTuneXML.element("uri",uri.toString());
+            }
+        return userTuneXML.toString();
+    }
+
+    private boolean isNullUserTune() {
+        if (this.artist == null && this.length == 0 && this.rating == 0 && this.source == null && this.title == null
+                && this.track == 0 && this.uri == null) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    public static boolean hasUserTuneElement(Message message) {
+        return message.hasExtension(ELEMENT, NAMESPACE);
+    }
+
+    public static UserTuneElement fromMessage(Message message) {
+        return message.getExtension(ELEMENT, NAMESPACE);
+    }
+
+    public UserTuneElement getUserTune() {
+        return this;
+    }
+
+    public static Builder getBuilder() {
+        return new Builder();
+    }
+    /**
+     * This class defines a Builder class for {@link UserTuneElement}.
+     * <br>
+     * {@link UserTuneElement} instance can be obtained using the {@link #build()} method as follows.
+     * <br>
+     * UserTuneElement.Builder builder = new UserTuneElement.Builder();
+     * builder.setSource("Yessongs"); builder.setTitle("Heart of the Sunrise");
+     * UserTuneElement userTuneElement = builder.build();
+     * <br>
+     * Values such as title, source, artist, length, source, track and uri can be set using their respective setters through {@link Builder}
+     * <br>
+     * Use {@link #buildEmptyUserTune()} only while disabling UserTune publish through {@link UserTuneManager#clearUserTune()}
+     */
+    public static final class Builder {
+        private String artist;
+        private int length = -1;
+        private int rating = -1;
+        private String source;
+        private String title;
+        private int track = -1;
+        private URI uri;
+
+        private Builder() {
+        }
+
+        public Builder setArtist(String artist) {
+            this.artist = artist;
+            return this;
+        }
+
+        public Builder setLength(int length) {
```

Comment:
You do not prevent setting negative lengths here.

Reply:
Negative lengths managed.